### ISSUE TITLE
Auto-instrumentation does not support read-only systems

### ISSUE THREAD

**Issue by asfaltboy** [DataDog/dd-trace-py#6023]
### Summary of problem

We are experiencing the same issue as described in #5291, but also in latest version 1.13.3.

From initial investigation, I believe it's because we use `securityContext.readOnlyRootFilesystem: true` - [reference](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/), tl;dr it prevents writes to the container's file system on mount. (see more details below)

I'm wondering if there's a way to add ddtrace packages to path without needing to call pip in a read-only system. I noticed that dd-trace-js does this by copying node_modules, equiv to python libs, to the mount path ([source](https://github.com/DataDog/dd-trace-js/blob/d638b4b7bc0eaee1be0044c181619378a41e069c/lib-injection/copy-lib.sh#L2)). This path is later added to the [require](https://nodejs.org/api/cli.html#-r---require-module) config key of NODE_OPTIONS in the main container ([source](https://github.com/DataDog/datadog-agent/blob/28f3540a85018c960d369b15eb3adcaf1a32b967/pkg/clusteragent/admission/mutate/auto_instrumentation.go#L37-L38)). Note: we use JS pods with securityContext and dd-trace-js works fine, because it only requires read access).

#### Simple solution proposal

As a proposed solution, how about installing the packages into `/datadog-lib` directory which is a volume we have control over, and can make read-write? This path is then added to the PYTHONPATH, which allows [the `sitecustomize.py` bootstrap script](https://github.com/DataDog/dd-trace-py/blob/v1.13.4/lib-injection/sitecustomize.py) to be called.

I feel this is a good first step to support this pod configuration; even though this still means granting write access on a path that is being executed (by python runtime), which is still an avenue that an attacker could use.

#### Better (but more complex) solution

An even better next step would be to install the wheels in `/datadog-lib/ddtrace_pkgs/` into `/datadog-lib/` _in the init container_ itself, which can be done at build time. The limitation of this is that we will need to ensure that all package sources are architecture agnotstic (no ABI restrictions), i.e include all `-py3-none-any.whl` or `source.tgz` packages (currently missing for some). Note that this could potentially mean a performance degredation if the package uses optimisations, for example something like a [CFFI](https://cffi.readthedocs.io/en/latest/) to an arch-specific precompiled lib.

What do you think, would that work?

### Which version of dd-trace-py are you using?

1.13.3.

### Which version of pip are you using?

23.1.2

### Which libraries and their versions are you using?

<details>
  <summary>`pip freeze`</summary>

```
aiohttp==3.8.4
aiosignal==1.3.1
alembic==1.11.1
anyio==3.6.2
async-timeout==4.0.2
asyncpg==0.27.0
attrs==23.1.0
backoff==2.2.1
bytecode==0.14.2
cattrs==22.2.0
certifi==2023.5.7
cffi==1.15.1
chardet==5.1.0
charset-normalizer==3.1.0
click==8.1.3
colorlog==6.7.0
cryptography==40.0.2
ddsketch==2.0.4
ddtrace==1.13.4
Deprecated==1.2.14
envier==0.4.0
fastapi==0.95.2
frozenlist==1.3.3
gcloud-aio-auth==4.2.1
gcloud-aio-pubsub==5.4.0
greenlet==2.0.2
h11==0.14.0
httpcore==0.17.1
httpx==0.24.1
idna==3.4
importlib-metadata==6.0.1
jsonschema==4.17.3
Mako==1.2.4
MarkupSafe==2.1.2
multidict==6.0.4
opentelemetry-api==1.18.0
packaging==23.1
prometheus-client==0.16.0
protobuf==4.23.2
pycparser==2.21
pydantic==1.10.8
PyJWT==2.7.0
pyrsistent==0.19.3
python-dotenv==1.0.0
python-json-logger==2.0.7
python-ulid==1.1.0
six==1.16.0
sniffio==1.3.0
SQLAlchemy==2.0.15
starlette==0.27.0
tenacity==8.2.2
typing_extensions==4.5.0
uvicorn==0.22.0
wrapt==1.15.0
xmltodict==0.13.0
yarl==1.9.2
zipp==3.15.0
```

</details>

### How can we reproduce your problem?

Set the security context of your pod spec, e.g:

<details>
  <summary>Click to expand!</summary>

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: security-context-demo
spec:
  securityContext:
    runAsUser: 1000
    runAsGroup: 3000
    fsGroup: 2000
  volumes:
  - name: sec-ctx-vol
    emptyDir: {}
  containers:
  - name: sec-ctx-demo
    image: busybox:1.28
    command: [ "sh", "-c", "sleep 1h" ]
    volumeMounts:
    - name: sec-ctx-vol
      mountPath: /data/demo
    securityContext:
      allowPrivilegeEscalation: false
      readOnlyRootFilesystem: true
```

</details>

See more info on how and why users would want to do this here:
* https://snyk.io/blog/10-kubernetes-security-context-settings-you-should-understand/
* https://www.fairwinds.com/blog/dont-forget-to-check-kubernetes-pod-securitycontext-for-readonlyrootfilesystem

### What is the result that you get?

Datadog bootstrap fails with:

```
ERROR: Could not install packages due to an OSError: [Errno 30] Read-only file system: '/code/.venv/lib/python3.11/site-packages/zipp'
```

<details>
  <summary>Click to see full log</summary>

```python
2023-05-27T11:37:37.797	WARNING: The directory '/home/user/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you should use sudo's -H flag.
2023-05-27T11:37:37.885	Looking in links: /datadog-lib/ddtrace_pkgs
2023-05-27T11:37:39.494	Collecting ddtrace
2023-05-27T11:37:39.545	  Downloading ddtrace-1.13.4-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.2 MB)
2023-05-27T11:37:39.601	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.2/4.2 MB 78.8 MB/s eta 0:00:00
2023-05-27T11:37:39.658	Collecting ddsketch>=2.0.1 (from ddtrace)
2023-05-27T11:37:39.664	  Downloading ddsketch-2.0.4-py3-none-any.whl (18 kB)
2023-05-27T11:37:39.695	Collecting packaging>=17.1 (from ddtrace)
2023-05-27T11:37:39.704	  Downloading packaging-23.1-py3-none-any.whl (48 kB)
2023-05-27T11:37:39.707	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 48.9/48.9 kB 81.6 MB/s eta 0:00:00
2023-05-27T11:37:39.733	Collecting tenacity>=5 (from ddtrace)
2023-05-27T11:37:39.749	  Downloading tenacity-8.2.2-py3-none-any.whl (24 kB)
2023-05-27T11:37:39.777	Collecting cattrs (from ddtrace)
2023-05-27T11:37:39.783	  Downloading cattrs-22.2.0-py3-none-any.whl (35 kB)
2023-05-27T11:37:39.806	Collecting six>=1.12.0 (from ddtrace)
2023-05-27T11:37:39.812	  Downloading six-1.16.0-py2.py3-none-any.whl (11 kB)
2023-05-27T11:37:39.814	Requirement already satisfied: typing-extensions in ./.venv/lib/python3.11/site-packages (from ddtrace) (4.5.0)
2023-05-27T11:37:39.850	Collecting jsonschema (from ddtrace)
2023-05-27T11:37:39.857	  Downloading jsonschema-4.17.3-py3-none-any.whl (90 kB)
2023-05-27T11:37:39.860	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 90.4/90.4 kB 85.0 MB/s eta 0:00:00
2023-05-27T11:37:39.881	Collecting xmltodict>=0.12 (from ddtrace)
2023-05-27T11:37:39.887	  Downloading xmltodict-0.13.0-py2.py3-none-any.whl (10.0 kB)
2023-05-27T11:37:39.905	Collecting envier (from ddtrace)
2023-05-27T11:37:39.911	  Downloading envier-0.4.0-py3-none-any.whl (10 kB)
2023-05-27T11:37:39.915	Requirement already satisfied: attrs>=20 in ./.venv/lib/python3.11/site-packages (from ddtrace) (23.1.0)
2023-05-27T11:37:40.141	Collecting protobuf>=3 (from ddtrace)
2023-05-27T11:37:40.148	  Downloading protobuf-4.23.2-cp37-abi3-manylinux2014_x86_64.whl (304 kB)
2023-05-27T11:37:40.153	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 304.5/304.5 kB 101.2 MB/s eta 0:00:00
2023-05-27T11:37:40.186	Collecting opentelemetry-api>=1 (from ddtrace)
2023-05-27T11:37:40.192	  Downloading opentelemetry_api-1.18.0-py3-none-any.whl (57 kB)
2023-05-27T11:37:40.195	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.3/57.3 kB 88.6 MB/s eta 0:00:00
2023-05-27T11:37:40.227	Collecting bytecode (from ddtrace)
2023-05-27T11:37:40.233	  Downloading bytecode-0.14.2-py3-none-any.whl (38 kB)
2023-05-27T11:37:40.296	Collecting deprecated>=1.2.6 (from opentelemetry-api>=1->ddtrace)
2023-05-27T11:37:40.303	  Downloading Deprecated-1.2.13-py2.py3-none-any.whl (9.6 kB)
2023-05-27T11:37:40.362	Collecting importlib-metadata~=6.0.0 (from opentelemetry-api>=1->ddtrace)
2023-05-27T11:37:40.368	  Downloading importlib_metadata-6.0.1-py3-none-any.whl (21 kB)
2023-05-27T11:37:40.371	Requirement already satisfied: setuptools>=16.0 in ./.venv/lib/python3.11/site-packages (from opentelemetry-api>=1->ddtrace) (66.1.1)
2023-05-27T11:37:40.500	Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0 (from jsonschema->ddtrace)
2023-05-27T11:37:40.507	  Downloading pyrsistent-0.19.3-py3-none-any.whl (57 kB)
2023-05-27T11:37:40.510	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.5/57.5 kB 83.6 MB/s eta 0:00:00
2023-05-27T11:37:40.640	Collecting wrapt<2,>=1.10 (from deprecated>=1.2.6->opentelemetry-api>=1->ddtrace)
2023-05-27T11:37:40.646	  Downloading wrapt-1.15.0-cp311-cp311-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (78 kB)
2023-05-27T11:37:40.650	     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.9/78.9 kB 75.9 MB/s eta 0:00:00
2023-05-27T11:37:40.705	Collecting zipp>=0.5 (from importlib-metadata~=6.0.0->opentelemetry-api>=1->ddtrace)
2023-05-27T11:37:40.712	  Downloading zipp-3.15.0-py3-none-any.whl (6.8 kB)
2023-05-27T11:37:41.159	Installing collected packages: zipp, xmltodict, wrapt, tenacity, six, pyrsistent, protobuf, packaging, envier, cattrs, bytecode, jsonschema, importlib-metadata, deprecated, ddsketch, opentelemetry-api, ddtrace
2023-05-27T11:37:41.161	ERROR: Could not install packages due to an OSError: [Errno 30] Read-only file system: '/code/.venv/lib/python3.11/site-packages/zipp'
2023-05-27T11:37:41.681	datadog autoinstrumentation: installing python package
2023-05-27T11:37:41.681	datadog autoinstrumentation: failed to install python package Command '['/code/.venv/bin/python', '-m', 'pip', 'install', '--find-links', '/datadog-lib/ddtrace_pkgs', 'ddtrace']' returned non-zero exit status 1.
```

</details>

### What is the result that you expected?

The instrumentation should _just work_ ™️

**Comment by emmettbutler** [comment]
Thank you for suggesting this improvement @asfaltboy. We'll take a look and evaluate how we might integrate it into the library.

**Comment by emmettbutler** [comment]
@asfaltboy This appears to have been fixed in 1.14.0. The fix is available in [this image](https://github.com/DataDog/dd-trace-py/pkgs/container/dd-trace-py%2Fdd-lib-python-init/98401642?tag=453f5bb52a923b2784c1f2c8477907694b731e4d).

**Comment by asfaltboy** [comment]
@emmettbutler , I don't think it was.

It's my understanding that the call to `pip install` ([here](https://github.com/DataDog/dd-trace-py/blob/v1.14.0/lib-injection/sitecustomize.py#L41)) is going to try and write to the python's `site-packages` directory, which in our case is located on the Read-only container file system.

In order to implement the first proposed solution above, consider using [the `pip install -t` option](https://pip.pypa.io/en/stable/cli/pip_install/#cmdoption-t) in the same pip call above.

Also, I appologise if I am missing something, please add more detail on how the issue described was fixed and if there's any actions we should follow?

**Comment by emmettbutler** [comment]
My mistake - sorry about that! I've reopened to keep track of the fix.

**Comment by asfaltboy** [comment]
@emmettbutler it seems @Kyle-Verhoog has fixed this the issue in #6047 which was merged last week. Happy to see that the solution he went with supports multiple platforms 🎉  and it seems it's already landed as a backport in 1.17 👏
